### PR TITLE
add stub test script

### PR DESCRIPTION
**Pull request by Andriamanitra**

### CODE REVIEW

File: scripts/teststub.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    clash_ids = []
+    for full_clash_path in glob.glob(f"{CLASH_DIR}/*.json"):
+        filename = basename(full_clash_path)
+        clash_id, _ = splitext(filename)
+        clash_ids.append(clash_id)
+
+    if n is not None:
+        clash_ids = random.sample(clash_ids, n)
+
+    GCC_CMD = ["gcc", "-fsyntax-only", "-x", "c", "-"]
```
Comment (ellnix) [inline scripts/teststub.py]: This could probably be a map of arrays? If every language has a single command.

**Review by ellnix — COMMENTED** [review]
The script looks good, There could be refactoring to make it more readable but I don't think that's worth doing at a file this size. It's perfectly understandable even for a noob like me.

I think the thing you could improve is to make this multi threaded. I don't know how multi threading works in python, but if it's anything like Rust it should be easy to partition the clashes in `thread_count` places and then run the `check_stubgen` command for each lang in each thread with their own subsample of clashes. That should also avoid having all the "expensive" languages in one slow thread.

File: scripts/teststub.py (hunk 1 above)
Comment (Andriamanitra) [reply scripts/teststub.py]: That's a great idea, I put the languages into a dict so it's easy to comment out or add languages

**Comment by daxida** [comment]
Is this something intended to be run for every language where we support a template? I just vaguely recall using it only for C and I'm surprised to see that loop over three languages.

How fast it fares for both of three if you happen to remember?

**Comment by Andriamanitra** [comment]
> Is this something intended to be run for every language where we support a template? I just vaguely recall using it only for C and I'm surprised to see that loop over three languages.

Nah it doesn't need to support every language, just the ones that are easy to check. I skipped interpreted languages for now because I don't really want to actually run the code. Maybe there's some static tooling that could help with those? The more we can test the better.

> How fast it fares for both of three if you happen to remember?

I've just left it running in the background for a while so I don't know how long it took exactly, but with these 3 languages it does about 5 clashes per second so that should be around 10 minutes for all 2776 clashes. I don't expect it to be ran very often so I'm not too worried about the speed.

**Pull request merged**